project-koku/koku
# COST-325: Fix multiple group_by org_unit_id for cost report.

#### Problem breakdown 

Jira Issue: https://issues.redhat.com/browse/COST-325

Currently there isn't a good way to do multiple group_by for both `group_by[or:org_unit_id]` & `group_by[org_unit_id]`. There wasn't really a solution for `group_by[or:org_unit_id]` because the key comes in as `or:org_unit_id` which means this check was never being hit: https://github.com/project-koku/koku/blob/master/koku/api/report/aws/query_handler.py#L186
```
if "org_unit_id" in self.parameters.parameters.get("group_by"):
            org_unit_applied = True
            # remove the org unit and add in group by account
            org_unit_group_by_data = self.parameters.parameters.get("group_by").pop("org_unit_id")
```

As for multiple group_bys for `group_by[org_unit_id]`, even though it would hit the condition we would only truly look at the first element in the list here: https://github.com/project-koku/koku/blob/master/koku/api/report/aws/query_handler.py#L197
```
org_unit_object = (
   AWSOrganizationalUnit.objects.filter(org_unit_id=org_unit_group_by_data[0])
   .filter(account_alias__isnull=True)
   .first()
)
```

The overall struggle for this issue was ultimately deciding what each multiple group by should look like in the return. After a lengthy discussion with Ash, we decided that the `group_by[or:org_unit_id]=OU_001&group_by[or:org_unit_id]=OU_002` would return the sub orgs 1 level deep for both OU_001 & OU_002. 

#### How to test
Here is a little diagram of the current tree structure:  https://docs.google.com/presentation/d/1Oa2njqavJVm7PJc_h8wqFx9uBrO2mER7WFzbPBG_-OQ/edit?usp=sharing

Run `make load-aws-org-unit-tree` to populate the tree in the database.

Hit the multiple `group_by[or:org_unit_id]` endpoint: 
- http://127.0.0.1:8000/api/cost-management/v1/reports/aws/costs/?delta=cost&filter[limit]=10&filter[offset]=0&filter[resolution]=monthly&filter[time_scope_units]=month&filter[time_scope_value]=-1&group_by[or:org_unit_id]=OU_001&group_by[or:org_unit_id]=OU_002&order_by[cost]=desc
- We need to make sure that the sub ous for `OU_001` and `OU_002` are shown in the results. Per the diagram above that would be `OU_003` & `OU_005`.

We also need to check that the single group by still works:
- http://127.0.0.1:8000/api/cost-management/v1/reports/aws/costs/?group_by[org_unit_id]=R_001, You should still see `OU_001` & `OU_002` in the returns.
- http://127.0.0.1:8000/api/cost-management/v1/reports/aws/costs/?group_by[oorg_unit_id]=OU_001&group_by[or:region]=*
- http://127.0.0.1:8000/api/cost-management/v1/reports/aws/costs/?group_by[org_unit_id]=OU_001&group_by[or:account]=*



The bad combinations caught in the serializer:
-  http://127.0.0.1:8000/api/cost-management/v1/reports/aws/costs/?group_by[org_unit_id]=OU_001&group_by[or:org_unit_id]=OU_002
- http://127.0.0.1:8000/api/cost-management/v1/reports/aws/costs/?group_by[org_unit_id]=OU_001&group_by[org_unit_id]=OU_002
- http://127.0.0.1:8000/api/cost-management/v1/reports/aws/costs/?group_by[or:org_unit_id]=*
